_posts/2017-01-17-http-keep-alive.markdown

```diff
@@ -0,0 +1,142 @@
+
+Looking into the Node documentation for the http/https module (where we should have started in the first place - RTFM right?) we find some interesting options for managing our connections. In particular, the [http.Agent](https://nodejs.org/api/http.html#http_class_http_agent) class was standing out. By default, http requests from Node can, but are not required to, have a custom `Agent` property on the `options` argument for an `http.request`. By omitting your own `Agent` option we were accepting the defaults. The answer was staring us in the face.
+
+From the `Agent` documentation:
+```
+const http = require('http');
+var keepAliveAgent = new http.Agent({ keepAlive: true });
+options.agent = keepAliveAgent;
+http.request(options, onResponseCallback);
+```
+
+The key here is the `{ keepAlive: true }`. As described by the documentation: *"Keep sockets around in a pool to be used by other requests in the future"*.
+
+Deploying this change was the solution to our problem. I was still bothered by the fact that our Node app wouldn't retry a request if it received an initial failure and that it would try for (by default with Heroku) 30 seconds before the connection was killed. There were certainly times where a connection would go astray on the first try and work perfectly well on the second. We don't want to wait around for 30 seconds to find out that we should have just tried a second time. To address this concern, we added a `timeout` option to our `http.request` options and then in our response processing, we looked for timeouts and issued a retry.
+
+Our node application uses the `isomorphic-fetch` package, which is an abstraction layer on top of the `http.request` module. So, here is an approximation of our code (shortened and simplified):
+
+```
+let response;
+let statusCode;
+const fetchOptions = {
+  agent: new http.Agent({ keepAlive: true }),
+  timeout: 5000,
+};
+
+...setup code for our request...
+
+response = await fetch(url, fetchOptions).catch(catchError);
+statusCode = response.status;
+
+...other code looking for errors...
+
+// retry once for timeouts
+if (statusCode === 503) {
+  console.error('RETRY', url);
+  response = await fetch(url, fetchOptions).catch(catchError);
+  statusCode = response.status;
+}
+```
+
+With this new code, we were now able to pool our sockets which fixed the timeout issues and we were able to retry our requests in the case of a temporary blip. In the end we wish that we had addressed these issues prior to launch but the troubleshooting process was a great learning experience. In my years as a software engineer I have learned that no matter diligent you are in your process and review you can't catch everything. The key is to learn from your mistakes and improve for next time.
```

Comment:
I think you missed a word.

`In my years as a software engineer I have learned that no matter` **how** `diligent you are in your process and review you can't catch everything.`